src/java/org/apache/cassandra/cql3/Vectors.java

```diff
@@ -0,0 +1,231 @@
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+public class Vectors
+{
+    private Vectors() {}
+
+    private static AbstractType<?> elementsType(AbstractType<?> type)
+    {
+        return ((VectorType<?>) type.unwrap()).getElementsType();
+    }
+
+    private static ColumnSpecification valueSpecOf(ColumnSpecification column)
+    {
+        return new ColumnSpecification(column.ksName, column.cfName, new ColumnIdentifier("value(" + column.name + ")", true), elementsType(column.type));
+    }
+
+    /**
+     * Tests that the list with the specified elements can be assigned to the specified column.
+     *
+     * @param receiver the receiving column
+     * @param elements the list elements
+     */
+    public static AssignmentTestable.TestResult testVectorAssignment(ColumnSpecification receiver,
+                                                                     List<? extends AssignmentTestable> elements)
+    {
+        if (!(receiver.type instanceof VectorType))
+            return AssignmentTestable.TestResult.NOT_ASSIGNABLE;
+
+        // If there is no elements, we can't say it's an exact match (an empty list if fundamentally polymorphic).
+        if (elements.isEmpty())
+            return AssignmentTestable.TestResult.WEAKLY_ASSIGNABLE;
+
+        ColumnSpecification valueSpec = valueSpecOf(receiver);
+        return AssignmentTestable.TestResult.testAll(receiver.ksName, valueSpec, elements);
+    }
+
+    /**
+     * Returns the exact ListType from the items if it can be known.
+     *
+     * @param items the items mapped to the list elements
+     * @param mapper the mapper used to retrieve the element types from the items
+     * @return the exact ListType from the items if it can be known or <code>null</code>
+     */
+    public static <T> VectorType<?> getExactListTypeIfKnown(List<T> items,
+                                                            java.util.function.Function<T, AbstractType<?>> mapper)
+    {
+        // TODO - this doesn't feel right... if you are dealing with a literal then the value is `null`, so we will ignore
+        // if there are multiple times, we randomly select the first?  This logic matches Lists.getExactListTypeIfKnown but feels flawed
+        Optional<AbstractType<?>> type = items.stream().map(mapper).filter(Objects::nonNull).findFirst();
+        return type.isPresent() ? VectorType.getInstance(type.get(), items.size()) : null;
+    }
+
+    public static <T> VectorType<?> getPreferredCompatibleType(List<T> items,
+                                                               java.util.function.Function<T, AbstractType<?>> mapper)
+    {
+        // TODO - this doesn't feel right... if you are dealing with a literal then the value is `null`, so we will ignore
+        // if there are multiple times, we randomly select the first?  This logic matches Lists.getExactListTypeIfKnown but feels flawed
```

Comment:
I guess it would be:
```suggestion
        // if there are multiple items, we randomly select the first?  This logic matches Lists.getPreferredCompatibleType but feels flawed
```
But I think that `AssignmentTestable#getCompatibleTypeIfKnown` selects an element compatible type that is also compatible with all the compatible types of all the other elements. So it doesn't randomly select the first type.

Reply:
yeah, for `getCompatibleTypeIfKnown` that is true, my comment was copy/paste from my comment in `getExactListTypeIfKnown`... the comment isn't valid for `getCompatibleTypeIfKnown`